### ISSUE TITLE
Testing asyncio application

### ISSUE THREAD

**Issue by drmikecrowe** [behave/behave#907]
Is it possible to test an application based on `asyncio` events?

My application is based on [pyee](https://pyee.readthedocs.io) using an event-bus to communicate between systems and [sismic](https://sismic.readthedocs.io/en/latest/) as the state machine. 

In my application, I start it using `aiorun` as follows:

```py
from aiorun import run
async def main():
    app = TestApplication()
    await shutdown_waits_for(app.cleanup())

run(main())
```

In this case, `TestApplication` is a testing stub that is doing the event routing between the state machine and various sub-systems listening to the event bus.  

What I'm trying to do is test the communication between systems.  To run, the `GuiApplication` has to be running in the event loop, so a give step definition (that isn't working) would be something like this:

```py
@when('I wait for message {message} to be fired containing {key}="{value}"')
@then('message {message} is fired containing {key}="{value}"')
@async_run_until_complete(timeout=1.2)
async def step_impl5(context, message, key=None, value=None):
    dict = {}
    if key:
        dict[key] = value
    ta = TestApplication()
    await ta.wait_for_message(message, dict)
```

However, the above terminates with `RuntimeError: This event loop is already running`

I've has a bit of success using [nest_asyncio](https://github.com/erdewit/nest_asyncio):

```py
import nest_asyncio
nest_asyncio.apply()
```

but this doesn't seem right either.

In short, how can I utilize the existing event loop for testing, or can I?

**Comment by jenisys** [comment]
Look at [behave/api/async_step.py](https://github.com/behave/behave/blob/master/behave/api/async_step.py).
Most/All constructs to support the async execution of steps have a `loop` parameter (default: `None`).
Therefore, you should be able to pass your event-loop to the decrotar(s) / class(es) / function(s).

NOTE:
There is another parameter `async_context` (with a type of `AsyncContext` class) that normally stores the event-loop parameter internally. This allows you to access (and potentially modify) the `loop` parameter within a step implementation (if necessary).

SEE ALSO:
* [behave/api/async_step.py](https://github.com/behave/behave/blob/master/behave/api/async_step.py)
* [features/step.async_steps.feature](https://github.com/behave/behave/blob/master/features/step.async_steps.feature)

**Comment by jenisys** [comment]
@drmikecrowe 
Design critique on mixture of `when`/ `then` steps.
`Then` steps are normally used to verify that a certain expected outcome occurred.
This is done by asserting that this condition is true (which is missing in your example).

**Comment by drmikecrowe** [comment]
@jenisys -- re: Design critique... Obvious in hindsight.  I must have realized that cuz they are gone now.

Regarding loops, I have created a LoopRunner following [this design pattern](https://github.com/nchazin/pycon2019/blob/master/presentation.pdf).  Each scenario, I'm starting another event-loop in a thread.  In the `after_scenario` step, I was going to `join` the thread and terminate the loop.  Any feedback on this at a high level?
